### PR TITLE
Add new_pos to moveresult output

### PR DESCRIPTION
**Pull request by oilboi**
Add compact, short information about your PR for easier understanding:

- Goal of the PR

To allow modders to move the entities back to the position where they collided.

- How does the PR work?

Adds a new table "new_pos" to the moveresult.collisions table.

Here is a video of it in action: https://youtu.be/g7NBuGFNKnU

- Does it resolve any reported issue?

https://github.com/minetest/minetest/issues/9966

## To do

Ready for Review.

## How to test

Download: https://github.com/oilboi/Crafter and shoot an arrow at a node at an angle.

Or create a mod called "debug_move" with this in the init.lua:

```
local test = {}
test.initial_properties = {
	physical = true,
	collide_with_objects = true,
	collisionbox = {-0.5, -0.5, -0.5, 0.5, 0.5, 0.5},
	visual = "cube",
	visual_size = {x = 1 , y = 1},
	textures = {"default_dirt.png","default_dirt.png","default_dirt.png","default_dirt.png","default_dirt.png","default_dirt.png"},
}
test.on_activate = function(self, staticdata, dtime_s)
    self.object:set_acceleration(vector.new(0,-9.81,0))
end
test.timer = 0
test.stuck = false
test.on_step = function(self, dtime,moveresult)
    if self.stuck == false and moveresult and moveresult.collides and moveresult.collisions and moveresult.collisions[1] and moveresult.collisions[1].new_pos then
        local movepos = moveresult.collisions[1].new_pos
        print("The object's new position (X:"..movepos.x.."|Y:"..movepos.y.."|Z:"..movepos.z..". Sticking!")
        self.object:set_velocity(vector.new(0,0,0))
        self.object:set_acceleration(vector.new(0,0,0))
        self.object:set_pos(moveresult.collisions[1].new_pos)
        self.stuck = true
    elseif self.stuck == true then
        self.timer = self.timer + dtime
        if self.timer > 3 then
            self.object:remove()
        end
    end
end
minetest.register_entity("debug_move:test", test)
minetest.register_chatcommand("spawnit", {
	params = "",
	description = "",
	privs = {server = true},
    func = function(name)
        local pos = minetest.get_player_by_name(name):get_pos()
        pos.y = pos.y + 3
        local vel = vector.new(math.random(-5,5),math.random(1,4),math.random(-5,5))
        local obj = minetest.add_entity(pos,"debug_move:test")
        if obj then
            obj:set_velocity(vel)
        end
    end
})
```

### CODE REVIEW

**Comment by KaylebJay** [comment]
Ahhh... this would make arrows attaching to players possible, then? :smile:

**Comment by oilboi** [comment]
Fixed the commit pollution, squash this when commited

**Comment by SmallJoker** [comment]
What's the difference to `object:get_pos()` at the time when it collides?

**Comment by sfan5** [comment]
Since [this loop](https://github.com/minetest/minetest/blob/master/src/collision.cpp#L445) runs multiple times, the object will have been moved somewhere else after a collision happens.

File: src/collision.cpp [hunk 1]
```diff
@@ -527,6 +527,10 @@ collisionMoveResult collisionMoveSimple(Environment *env, IGameDef *gamedef,
 				is_collision = false;
 
 			CollisionInfo info;
+			if (is_collision) {
+				info.new_pos = *pos_f/10;
```
Comment (sfan5) [inline src/collision.cpp:531]: should be `*pos_f / BS`
this can also be moved into the `if (is_collision)` on line 569

Comment (oilboi) [reply src/collision.cpp]: Updated!

**Comment by sfan5** [comment]
:+1: Tested and works.

I am wondering though if this could be better solved by something like `sticky_collision`, where the collision + movement stops after the first hit. This would also allow the client to predict the arrow properly.

Are there any usecases where you'd need `new_pos` but <i>don't</i> want the object to stick where it collided?

**Comment by SmallJoker** [comment]
A `stop_on_collide` object property might make more sense in this case?

**Comment by oilboi** [comment]
That could be built on top of this PR once merged :D

File: src/collision.cpp [hunk 2]
```diff
@@ -527,6 +527,10 @@ collisionMoveResult collisionMoveSimple(Environment *env, IGameDef *gamedef,
 				is_collision = false;
 
 			CollisionInfo info;
+			if (is_collision) {
+				info.new_pos = *pos_f / BS;
```
Comment (SmallJoker) [inline src/collision.cpp:531]: Division should be moved to `push_collision_move_result`. Please don't mix Minetest internals (factor `BS`) with exposed values.

**Comment by sorcerykid** [comment]
This seems much too hacky to me. Adding a feature so that developers can counteract an unwanted behavior of the engine is not how an API should be designed. I think it would be more sensible to have a property to prevent the object's position being changed when a collision occurs as suggested by others.

**Comment by sfan5** [comment]
The other side of that argument is that it's better to give modders the tools to something imperfectly and add engine support if it is deemed necessary later.
E.g. I can't directly think of an usecase for the moveresult `old_speed` / `new_speed`, yet it was useful for the builtin falling nodes to implement this workaround:
https://github.com/minetest/minetest/blob/7ec0e3df35a11b66b48eababf0123170f2453a50/builtin/game/falling.lua#L314-L327

**Comment by sorcerykid** [comment]
That seems to contravenes your earlier argument where you opposed adding texture modifiers as an interim workaround for on-the-fly texture creation, claiming it's a bad approach and supposedly a "hack", even though I can see many use-cases for programmatically generated PNG textures.  

> Modders often want to send textures at runtime, which has previously used hacks like [combine to build a texture pixel by pixel.
Other suggested approaches are #9767, which packs pixels into a texture modifier (also bad) or adding a texture modifier that can read encoded PNG data. 

I esp. thought this statement was telling: "The reason I dislike those approaches is that these workarounds can never be removed even if a proper solution arrives one day."

It doesn't seem very consistent that sometimes hacky APIs to workaround engine shortcomings that can never be removed are acceptable, but other times not.

**Comment by sfan5** [comment]
The difference to this is that `new_pos` has potential future applications, while a texture modifier that takes PNG data can be fully replaced (depends on the implementation, of course) by native engine functionality.
The other reason for my clear stance on #9767 is that the "pack texture into string" idea is outstandingly inefficient in multiple regards (caching, transfer size). In comparison adding a `new_pos` vector to moveresult doesn't really matter. \*

\* Doesn't mean I don't worry about the performance implications of pushing moveresult every step. This could be addressed in the future by perhaps adding a flag that says "do not tell me about my collision state unless I am/was moving".

**Comment by oilboi** [comment]
I cannot understand why the server reviewing a few more bytes of data for modders to utilize is creating an uproar.

**Comment by sorcerykid** [comment]
So far nobody has identified a use case for this additional field. All that has been said is that at some point it might be useful. To me that's not a strong argument for a change to the API no matter how many or few bytes it is. Multiple people have already proposed the optimal solution, which is for the object to "stick" when it collides.

**Comment by paramat** [comment]
IRC:
> 20:55 oil_boi    So I guess I should just close #9978?

There is absolutely no need to close this. It already has 1 core dev approval, and has no core dev disapprovals, so it is progressing well. It takes at least 2 core dev disapprovals to reject a PR.

sfan5
> Doesn't mean I don't worry about the performance implications of pushing moveresult every step.

Is this something the changed code does (i cannot see how), or what a modder would have to do to use it? What you wrote sounds concerning as if there is a significant performance hit from this code change.

**Comment by sfan5** [comment]
> Is this something the changed code does (i cannot see how), or what a modder would have to do to use it? What you wrote sounds concerning as if there is a significant performance hit from this code change.

The moveresult was being pushed every step already.
There is no significant performance difference with this PR, it's just that adding another property to the moveresult logically consumes more time than not doing so.

**Comment by oilboi** [comment]
Also: item entities can create a non-physical state by using the output of new_pos to stop them from moving

**Comment by LoneWolfHT** [comment]
Something like sticky_collision would from what I can tell cover everything new_pos will. But comparing the two with different usecases I think the modder's code will have the same performance hit in the end, the new code 'slowing things down' will just be in a different place

**Comment by LoneWolfHT** [comment]
@oilboi might want to close this PR before doing things with your engine fork

**Comment by sorcerykid** [comment]
How does that work exactly? Setting `sticky_collision` to true takes effectively zero CPU time and is executed once in the lifetime of an entity and allows for client-side preduction. That effectively reduces all of the code in the original post to nothing, and of course eliminates the extraneous API calls.

I implemented a variant of `sticky_collisions` on my server for use with bows and arrows in my Thievery mod, and it works flawlessly. The idea that modders should have to invest time coding all of those extra checks for a basic game mechanic is silly. APIs should not be designed with the goal of intentional obfuscation.

**Comment by LoneWolfHT** [comment]
In the case of my grenades mod:
If I wanted to make them bounce off of the exact place where they collided (Assuming they don't, I'd need to check) I would need to enable sticky_collision and run set_velocity() and set_acceleration() after every collision.
With new_pos I would need to set_pos() and set_velocity() in addition to a set_acceleration() (If I'm slowing my grenade's velocity down with that)

Another example might be bullets with chip particles to show where on a block they hit.
With sticky_collision I'd need to run get_pos(), remove the bullet, and spawn in a chip particle at that pos
With new_pos I'd need to remove the bullet and spawn in a chip particle at that pos

I can make a TNT cannon in vanilla MTG that launches falling nodes, if you wanted to make the pos where the falling entity becomes a node exact you could use sticky_position, but if you hit a wall you're going to have to set the velocity again. Unlike with new_pos where you can just ignore that collision if the falling node isn't touching the ground. (There are probably similar situations to this one too)

I haven't used moveresult yet so some of these situations might be nonexistent, let me know if I'm wrong
#
> The idea that modders should have to invest time coding all of those extra checks for a basic game mechanic is silly

This feature covers more than just one basic game mechanic though

> APIs should not be designed with the goal of intentional obfuscation.

I don't think the intentions of anyone here has anything to do with obfuscation.
I also don't think new_pos is all that confusing to use, and if you do think it's confusing then consider the situations I listed above, I'd say using sticky_collisions in some of those cases would be 'confusing' (Can't find the right word) too

**Pull request closed without merging**